### ISSUE TITLE
A few Priv_migrate issues

### ISSUE THREAD

**Issue by sn0wfa11** [rapid7/metasploit-framework#8035]
# Overview

Now that post/windows/manage/priv_migrate is the default behavior for Meterpreter "migrate -f", I wanted to document some issues that I have identified and provide a PR for the fix, along with a few improvements. 

**PR to follow shortly**

During a pentest class, I identified two issues with this module while interacting with a Windows Vista x86 machine. Both issues are repeatable with a standard Vista install.

## Steps to reproduce

1. Get a Windows Vista x86 VM set up. My test machine is as follows
  a. Windows Vista Enterprise x86 SP1
  b. Firewall on
  c. UAC on
  d. Internet Explorer 7
  e. Updates turned off
2. For the first part, use exploit/windows/browser/ie_execcommand_uaf and windows/meterpreter/reverse_tcp. Set up the server for the exploit and browse to it on the Vista machine for a low priv shell.
3. Use post/windows/manage/priv_migrate against that session. It will throw an unhandled error shown below when it cannot migrate into the spawned explorer.exe process.
```
[*] Meterpreter session 1 opened (10.8.0.66:13007 -> 192.168.1.218:49172) at 2017-02-28 18:57:45 -0600
[*] Session ID 1 (10.8.0.66:13007 -> 192.168.1.218:49172) processing AutoRunScript 'multi_console_command -rc /root/Desktop/migrate.rc'
[*] Running Command List ...
[*] 	Running command run post/windows/manage/priv_migrate
[*] Current session process is iexplore.exe (3740) as: WIN-0HORFV9LBZO\Josh
[*] Session has User level rights.
[*] Will attempt to migrate to a User level process.
[-] Could not migrate to explorer.exe.
[*] Attempting to spawn explorer.exe
[+] Successfully spawned explorer.exe
[*] Trying explorer.exe (908)
[-] Post failed: Rex::RuntimeError Cannot migrate into non existent process
[-] Call stack:
[-]   /usr/share/metasploit-framework/modules/post/windows/manage/priv_migrate.rb:105:in `migrate'
[-]   /usr/share/metasploit-framework/modules/post/windows/manage/priv_migrate.rb:163:in `block in migrate_user'
[-]   /usr/share/metasploit-framework/modules/post/windows/manage/priv_migrate.rb:157:in `each'
[-]   /usr/share/metasploit-framework/modules/post/windows/manage/priv_migrate.rb:157:in `migrate_user'
[-]   /usr/share/metasploit-framework/modules/post/windows/manage/priv_migrate.rb:51:in `run'
```

4. For the second issue: Use the low priv session and add a route to the target Vista VM.
5. Use exploit/windows/smb/ms09_050_smb2_negotiate_func_index against the VM now that the route provides access to port 445. Again use windows/meterpreter/reverse_tcp. Run the exploit for a System shell.
5. Run post/windows/manage/priv_migrate against the new System level session. It will fail to migrate into the system level processes and potentially downgrade the privileged shell, which it did in my class. However, the current test machine fails all migrates and the VM reboots.
```
[*] Meterpreter session 2 opened (10.8.0.66:13007 -> 192.168.1.218:49176) at 2017-02-28 19:04:21 -0600
[*] Session ID 2 (10.8.0.66:13007 -> 192.168.1.218:49176) processing AutoRunScript 'multi_console_command -rc /root/Desktop/migrate.rc'
[*] Running Command List ...
[*] 	Running command run post/windows/manage/priv_migrate
[*] Current session process is lsass.exe (692) as: NT AUTHORITY\SYSTEM
[*] Session is already Admin and System.
[*] Will attempt to migrate to specified System level process.
[*] Trying services.exe (640)
[-] Could not migrate to services.exe.
[-] stdapi_sys_config_getuid: Operation failed: 1722
[-] Could not migrate to winlogon.exe.
[-] Could not migrate to wininit.exe.
[-] Could not migrate to lsm.exe.
[-] Could not migrate to lsass.exe.
[-] Unable to migrate to any of the System level processes.
[*] Will attempt to migrate to a User level process.
[-] Could not migrate to explorer.exe.
[*] Attempting to spawn explorer.exe
[+] Successfully spawned explorer.exe
[*] Trying explorer.exe (2280)
[-] Could not migrate to explorer.exe.
[-] stdapi_sys_config_getuid: Operation failed: 1722
[-] Could not migrate to notepad.exe.
[*] Attempting to spawn notepad.exe
[+] Successfully spawned notepad.exe
[*] Trying notepad.exe (3416)
[-] Could not migrate to notepad.exe.
[-] stdapi_sys_config_getuid: Operation failed: 1722
```

## Expected behavior
1. It should not throw the unhandled error and quit. It should recover and move on. (easy fix)
2. The module should check if the session is already in one of the target processes and not try to migrate if that is the case.
3. The module should have an option to turn off the automatic user level migration on privileged shells if the admin migration fails. (More on this in the PR)

## Current behavior

See above.

### Metasploit version
Framework: 4.13.24-dev
Console  : 4.13.24-dev
and
Framework: 4.14.1-dev-def5088097
Console  : 4.14.1-dev-def5088097

### I installed Metasploit with:
- Kali package via apt
and
- Source install - Dev Environment (ruby version - 2.3.3)

### OS

Kali Rolling

**Comment by Pentoman** [comment]
Give the settings of the payload migrate priv itself please

it looks like u should use 

show options 

and put in your own parameters ...also some more delay and timout ??

make ps on the meterpeter 

to see the current processes then you can use one of them to migrate ...

**Comment by sn0wfa11** [comment]
Can you please clarify your comment? I'm unsure as to what you are asking or stating as it relates to the issues with the priv_migrate module.

**Comment by void-in** [comment]
@Pentoman The post module required datastore options are already filled with default values. When run automatically, the SESSION datastore option is populated with the invoking session so no need to configure anything. What @sn0wfa11 has identified here is related to some issues when the post module is already successfully invoked for the returned session but couldn't perform its intended purpose.

**Comment by sn0wfa11** [comment]
Fixed by PR #8037
